### PR TITLE
[Chore] Refactor Test Runs

### PR DESCRIPTION
**Pull request by anditurdiu**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
In #2738 we introduced a new boolean flag for the participation: `test_run`. This PR is a follow-up to refactor the test run feature, so that we only use this flag (together with `studentExam.test_run`) to distinguish test run student exams, participations, submissions, and results. Additionally, through the new changes, we can more closely align the test runs to real exams, making them more robust to changes and providing the instructors confidence that everything will work out during the exam. 

**The following issues are addressed in this PR:** 

**Server:** 

1. The repositories still use the old testRun identification condition: `result.assessor = participation.user` in most queries.
2. We do not need to mark the submissions of test runs anymore by adding a dummy result with the condition: `result.assessor = participation.user`
3. Test run quizzes are never evaluated and never display the users result.

**Client:** 
1. After the test run conduction, the results are immediately displayed leading to confusing notices, because there are no assessments yet, and quiz exercise information is hidden during the conduction. 
2. The second correction currently breaks the complaint functionality of test runs. (This happens because there is no second correction for test runs, and it tries to access a result which is not present)
3. In #2738 we delete all test runs, if the user clicks on "Assess all unsubmitted exams and empty submissions" after the exam conduction. This however is not mentioned to the user. 
4. The status indicator of the assessments per exercise is disabled for test runs, however the client still calls  `getStatsForTutors` unnecessarily.
5. When assessing the different test run participation for exercises, in the exercise assessment view the user cannot distinguish between real assessments and test run assessments. 
6. In the "create test run" modal, the user has to manually select every exercise for each exercise group, even if there is only one option.

### Description

- [x] Remove the previous testRun identification `result.assessor != participation.user` from all queries, and replace with `p.testRun = FALSE OR p.testRun IS NULL`
- [x] Remove the old `markSubmissionsOfTestRuns` method.
- [x] After submitting a test run, immediately evaluate the quiz participation (but ignore it from the quiz statistics).
- [x] Rework the test run submission assessment call to use the `lockSubmission` method of `SubmissionService` through a wrapper in order to not alter the code.
- [x] Only display the results of the test run when the user explicitly accesses the summary. _(similar to the real exam participation where the results are published at a later date, i.e, not immediately after the conduction.)_
- [x] Because there is no second correction for test runs, it does not make sense to display the "enable second correction button". Disable the button for test runs and display a notice to the user: "Second correction does not apply to test runs", if the exam has more than one correction round.
- [x] Adapt the notice to delete the test run in the "Test Run Management" view to let the instructor know that the test runs will be automatically deleted once he assesses all unsubmitted exams and empty submissions.
- [x] Do not call  `getStatsForTutors` for Test runs, as the status indicator is disabled.
- [x] Add "Test Run" to the header of the exercise assessment in order to help the user distinguish between assessing real submissions and test run submissions.
- [x] Automatically select the exercise for every exercise group, if there is only one in the "create test run" modal.  

**Unrelated changes:** 

- [x] Rename left-over "tutor-dashboard" calls to assessment-dashboard
- [x] Tutor leaderboards and side panels are disabled for exams. Therefore, do not load the `TutorLeaderboardEntries` (unneccesary database calls). 

**WARNING:** Deleting old test runs will fail because the `AssessmentType.TEST_RUN` was deleted from the server. In order for this to work, you must run the following query on your database:

> Update result set assessment_type = 'MANUAL' where assessment_type = 'TEST_RUN'

### Steps for Testing

**Test Runs**
1. Create an exam with all exercise types
2. Create a test run. If you only have one exercise per exercise group, it should be automatically selected. 
3. Change the working time vie Manage Working Time from the Test Run Management view.
4. Start the test run and submit. You should not see "No graded results", on top of exercises.
5. Assess the test run submissions. All assessment pages should include "Test Run" in the title.
6. Go to the summary page, you should see your results (including the quiz result)
7. Complain about the assessment.
8. Respond to the complaints.
9. View the complaints in the summary.
10. Change the exam to have 2 correction rounds.
11. Click on assess your test runs. It should now display: `Hint: The second correction does not apply to test runs.`
12. Go to the scores page, no test run should be included here.

**Normal exam**
_Verify that everything works as expected, there should be no changes in behaviour_
1. Register users for the exam and create the student exams & prepare exercises.
2. Change the working time of the student (make sure no console errors are thrown!).
3. Participate in the exam.
4. Assess all unsubmitted exams and empty submissions from the "student exams" view.
5. Go to Test Runs, and verify that all test runs were deleted. 
6. Assess the exam with 2 correction rounds enabled
7. Publish the results and view them 
8. Complain about the results
9. Respond to complaints
10. View the scores page

 

### Test Coverage
SubmissionResource: 91%
StudentExamResource: 92%

ExamService: 91%
StudentExamService: 90%
SubmissionService: 92%
ParticipationService: 79%
ExamQuizSubmissionService: 68%

### Screenshots
image
image
image
image

### CODE REVIEW

**Review by airelawaleria — DISMISSED** [review]
tested successfully on ts1

**Review by EneaGore — DISMISSED** [review]

**Review by JohannesWeiss — APPROVED** [review]
Worked as expected on ts1

**Review by marvinchimeka — DISMISSED** [review]
Tested with @simonzilker. No issues found. Works as intended.

**Review by simonzilker — DISMISSED** [review]
Tested with @marvinchimeka. Works as described!

File: src/main/java/de/tum/in/www1/artemis/repository/ComplaintResponseRepository.java [hunk 1]
```diff
@@ -46,15 +46,14 @@
     @Query("""
             SELECT COUNT (DISTINCT p) FROM StudentParticipation p
             WHERE p.exercise.id = :#{#exerciseId}
+            AND (p.testRun = FALSE OR p.testRun IS NULL)
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/repository/ComplaintResponseRepository.java:49]:
```suggestion
            AND p.testRun = FALSE
```

Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/repository/ComplaintResponseRepository.java:49]: the value cannot be null, and OR can be significantly slower in such cases

File: src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java [hunk 2]
```diff
@@ -134,7 +134,14 @@
      * @param exerciseId the exercise id we are interested in
      * @return the number of distinct submissions belonging to the exercise id
      */
-    @Query("SELECT COUNT (DISTINCT p) FROM ProgrammingExerciseStudentParticipation p WHERE p.exercise.id = :#{#exerciseId} AND EXISTS (SELECT s FROM ProgrammingSubmission s WHERE s.participation.id = p.id AND s.submitted = TRUE) AND NOT EXISTS (select prs from p.results prs where prs.assessor.id = p.student.id)")
+    @Query("""
+            SELECT COUNT (DISTINCT p) FROM ProgrammingExerciseStudentParticipation p
+            WHERE p.exercise.id = :#{#exerciseId}
+            AND (p.testRun = FALSE OR p.testRun IS NULL)
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java:140]:
```suggestion
            AND p.testRun = FALSE
```

Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java:140]: the value cannot be null, and OR can be significantly slower in such cases

File: src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java [hunk 3]
```diff
@@ -167,31 +174,31 @@ AND EXISTS (SELECT r.assessor FROM s.results r
     @Query("""
             SELECT COUNT (DISTINCT p) FROM ProgrammingExerciseStudentParticipation p
             WHERE p.exercise.id = :#{#exerciseId}
+            AND (p.testRun = FALSE OR p.testRun IS NULL)
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java:177]:
```suggestion
            AND p.testRun = FALSE
```

File: src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java [hunk 4]
```diff
@@ -167,31 +174,31 @@ AND EXISTS (SELECT r.assessor FROM s.results r
...
-                                AND r.submission.submitted = TRUE
-                                AND r.completionDate IS NOT NULL
-                                AND (p.exercise.dueDate IS NULL OR r.submission.submissionDate <= p.exercise.dueDate)
-                                AND NOT EXISTS (select prs from p.results prs where prs.assessor.id = p.student.id)
-                            ) >= (:correctionRound + 1L)
-            """)
+            SELECT COUNT(DISTINCT p)
+            FROM ProgrammingExerciseStudentParticipation p
+            WHERE p.exercise.id = :exerciseId
+            AND (p.testRun = FALSE OR p.testRun IS NULL)
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java:191]:
```suggestion
            AND p.testRun = FALSE
```

File: src/main/java/de/tum/in/www1/artemis/repository/ResultRepository.java [hunk 5]
```diff
@@ -96,7 +96,18 @@
     @Query("SELECT COUNT(DISTINCT p) FROM StudentParticipation p left join p.results r WHERE p.exercise.id = :exerciseId AND r.assessor IS NOT NULL AND r.rated = TRUE AND r.completionDate IS NOT NULL AND (p.exercise.dueDate IS NULL OR r.submission.submissionDate <= p.exercise.dueDate)")
     long countNumberOfFinishedAssessmentsForExercise(@Param("exerciseId") Long exerciseId);
 
-    @Query("SELECT COUNT(DISTINCT p) FROM StudentParticipation p left join p.results r WHERE p.exercise.id = :exerciseId AND r.assessor IS NOT NULL AND r.rated = TRUE AND r.submission.submitted = TRUE AND r.completionDate IS NOT NULL AND (p.exercise.dueDate IS NULL OR r.submission.submissionDate <= p.exercise.dueDate) AND NOT EXISTS (select prs from p.results prs where prs.assessor.id = p.student.id)")
+    @Query("""
+            SELECT COUNT(DISTINCT p) FROM StudentParticipation p
+            left join p.results r
+            WHERE p.exercise.id = :exerciseId
+            AND (p.testRun = FALSE OR p.testRun IS NULL)
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/repository/ResultRepository.java:103]:
```suggestion
            AND p.testRun = FALSE
```

File: src/main/java/de/tum/in/www1/artemis/repository/ResultRepository.java [hunk 6]
```diff
@@ -105,18 +116,18 @@
...
-                                AND r.rated = TRUE
-                                AND r.submission = (select max(id) from p.submissions)
-                                AND r.submission.submitted = TRUE
-                                AND r.completionDate IS NOT NULL
-                                AND (p.exercise.dueDate IS NULL OR r.submission.submissionDate <= p.exercise.dueDate)
-                                AND NOT EXISTS (select prs from p.results prs where prs.assessor.id = p.student.id)
-                            ) >= (:correctionRound + 1L)
+            SELECT COUNT(DISTINCT p)
+            FROM StudentParticipation p WHERE p.exercise.id = :exerciseId
+            AND (p.testRun = FALSE OR p.testRun IS NULL)
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/repository/ResultRepository.java:121]:
```suggestion
            AND p.testRun = FALSE
```

File: src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java [hunk 7]
```diff
@@ -208,9 +222,9 @@ and not exists (select prs from participation.results prs
     List<StudentParticipation> findAllWithEagerSubmissionsAndEagerResultsAndEagerAssessorByExerciseId(long exerciseId);
 
     @Query("""
-            SELECT DISTINCT p FROM StudentParticipation p left join fetch p.submissions s left join fetch s.results r left join fetch r.assessor a
+            SELECT DISTINCT p FROM StudentParticipation p LEFT JOIN FETCH p.submissions s LEFT JOIN FETCH s.results r LEFT JOIN FETCH r.assessor a
             WHERE p.exercise.id = :#{#exerciseId}
-            AND NOT EXISTS (select prs from p.results prs where prs.assessor.id = p.student.id)
+            AND (p.testRun = FALSE OR p.testRun IS NULL)
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java:227]:
```suggestion
            AND p.testRun = FALSE
```

File: src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java [hunk 8]
```diff
@@ -229,8 +243,13 @@ AND NOT EXISTS (select prs from p.results prs where prs.assessor.id = p.student.
     @Query("select distinct p from StudentParticipation p where p.student.id = :#{#studentId} and p.exercise in :#{#exercises}")
     List<StudentParticipation> findByStudentIdAndIndividualExercises(@Param("studentId") Long studentId, @Param("exercises") List<Exercise> exercises);
 
-    @Query("select distinct p from StudentParticipation p left join fetch p.submissions s left join fetch s.results r where p.student.id = :#{#studentId} and p.exercise in :#{#exercises}")
-    List<StudentParticipation> findByStudentIdAndIndividualExercisesWithEagerSubmissionsResult(@Param("studentId") Long studentId, @Param("exercises") List<Exercise> exercises);
+    @Query("select distinct p from StudentParticipation p left join fetch p.submissions s left join fetch s.results r where (p.testRun = false or p.testRun is null) and p.student.id = :#{#studentId} and p.exercise in :#{#exercises}")
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java:246]:
```suggestion
    @Query("select distinct p from StudentParticipation p left join fetch p.submissions s left join fetch s.results r where p.testRun = false and p.student.id = :#{#studentId} and p.exercise in :#{#exercises}")
```

File: src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java [hunk 9]
```diff
@@ -245,10 +264,10 @@ AND NOT EXISTS (select prs from p.results prs where prs.assessor.id = p.student.
     @Query("""
             select distinct p from StudentParticipation p left join fetch p.submissions s left join fetch s.results r
             where p.exercise.id = :#{#exerciseId}
+            and (p.testRun = FALSE OR p.testRun IS NULL)
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java:267]:
```suggestion
            and p.testRun = FALSE
```

File: src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java [hunk 10]
```diff
@@ -275,13 +294,12 @@ AND NOT EXISTS (select prs from p.results prs where prs.assessor.id = p.student.
     @Query("""
             SELECT distinct p FROM StudentParticipation p LEFT JOIN FETCH p.submissions s LEFT JOIN FETCH s.results r
                 WHERE p.exercise.id = :#{#exerciseId}
+                AND (p.testRun = FALSE OR p.testRun IS NULL)
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java:297]:
```suggestion
                AND p.testRun = FALSE
```

File: src/main/java/de/tum/in/www1/artemis/repository/SubmissionRepository.java [hunk 11]
```diff
@@ -105,7 +105,17 @@
      * @return the number of submissions belonging to the exercise id, which have the submitted flag set to true and the submission date before the exercise due date, or no
      *         exercise due date at all
      */
-    @Query("SELECT COUNT (DISTINCT p) FROM StudentParticipation p WHERE p.exercise.id = :#{#exerciseId} AND EXISTS (SELECT s FROM Submission s WHERE s.participation.id = p.id AND s.submitted = TRUE AND (p.exercise.dueDate IS NULL OR s.submissionDate <= p.exercise.dueDate)) AND NOT EXISTS (select prs from p.results prs where prs.assessor.id = p.student.id)")
+    @Query("""
+            SELECT COUNT (DISTINCT p) FROM StudentParticipation p
+            WHERE p.exercise.id = :#{#exerciseId}
+            AND (p.testRun = FALSE OR p.testRun IS NULL)
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/repository/SubmissionRepository.java:111]:
```suggestion
            AND p.testRun = FALSE
```

**Review by krusche — CHANGES_REQUESTED** [review]
In general the code changes look good, even if I did not review those in detail.
I have some inline comments though.

**Important:** We should test the changes really well to avoid issues!

**Review by tiffca — DISMISSED** [review]
All worked well except the last step, the complaint didn't show up for me. It looks like an isolated behavior, hence approving. 

image

File: src/main/java/de/tum/in/www1/artemis/repository/ComplaintResponseRepository.java (hunk 1 above)
Comment (anditurdiu) [reply src/main/java/de/tum/in/www1/artemis/repository/ComplaintResponseRepository.java]: fixed in 139c8fb547deadc2b10c0e497d91389dcb5f736c

File: src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java (hunk 2 above)
Comment (anditurdiu) [reply src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java]: fixed in 139c8fb547deadc2b10c0e497d91389dcb5f736c

File: src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java (hunk 3 above)
Comment (anditurdiu) [reply src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java]: fixed in 139c8fb547deadc2b10c0e497d91389dcb5f736c

File: src/main/java/de/tum/in/www1/artemis/repository/ResultRepository.java (hunk 5 above)
Comment (anditurdiu) [reply src/main/java/de/tum/in/www1/artemis/repository/ResultRepository.java]: fixed in 139c8fb547deadc2b10c0e497d91389dcb5f736c

File: src/main/java/de/tum/in/www1/artemis/repository/ResultRepository.java (hunk 6 above)
Comment (anditurdiu) [reply src/main/java/de/tum/in/www1/artemis/repository/ResultRepository.java]: fixed in 139c8fb547deadc2b10c0e497d91389dcb5f736c

File: src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java (hunk 7 above)
Comment (anditurdiu) [reply src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java]: fixed in 139c8fb547deadc2b10c0e497d91389dcb5f736c

File: src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java (hunk 8 above)
Comment (anditurdiu) [reply src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java]: fixed in 139c8fb547deadc2b10c0e497d91389dcb5f736c

File: src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java (hunk 9 above)
Comment (anditurdiu) [reply src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java]: fixed in 139c8fb547deadc2b10c0e497d91389dcb5f736c

File: src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java (hunk 10 above)
Comment (anditurdiu) [reply src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java]: fixed in 139c8fb547deadc2b10c0e497d91389dcb5f736c

File: src/main/java/de/tum/in/www1/artemis/repository/SubmissionRepository.java (hunk 11 above)
Comment (anditurdiu) [reply src/main/java/de/tum/in/www1/artemis/repository/SubmissionRepository.java]: fixed in 139c8fb547deadc2b10c0e497d91389dcb5f736c

**Review by nicolasruscher — DISMISSED** [review]
The code looks good to me, but I did not have time to test it..

**Review by airelawaleria — APPROVED** [review]

**Review by JohannesWeiss — APPROVED** [review]

**Review by SimonEntholzer — DISMISSED** [review]
The code looks good. Tested on ts1, the test-runs work as described, so I guess everything is fine :)

File: src/main/java/de/tum/in/www1/artemis/service/ExamQuizService.java [hunk 12]
```diff
@@ -85,81 +135,84 @@ public void evaluateQuizAndUpdateStatistics(@NotNull Long quizExerciseId) {
...
+                    // NOTE: we save participation, submission and result here individually so that one exception (e.g. duplicated key) cannot destroy multiple student answers
+                    submissionRepository.save(quizSubmission);
+                    result = resultRepository.save(result);
+
+                    // add result to participation
+                    participation.addResult(result);
+                    studentParticipationRepository.save(participation);
+
+                    // add result to submission
+                    // quizSubmission.replaceLatestOrIfEmptyAddResult(result);
```
[the remark this answers is not in the document]
Comment (anditurdiu) [reply src/main/java/de/tum/in/www1/artemis/service/ExamQuizService.java:199]: This is already the case in develop, I did not change this code snippet, just moved it inside an if check for test runs. I guess we can probably delete this

[the remark this answers is not in the document]
Comment (anditurdiu) [reply src/main/java/de/tum/in/www1/artemis/service/ExamQuizService.java]: Removed in: 009f2f0

**Review by airelawaleria — DISMISSED** [review]

**Review by JohannesWeiss — APPROVED** [review]

**Review by SimonEntholzer — DISMISSED** [review]

**Review by philgabr — APPROVED** [review]

**Review by fde312 — COMMENTED** [review]
I just found a small issue. When you are not a instructor in the course but have admin rights, you cannot submit a complaint. After adding myself as instructor I was able to.

**Review by SimonEntholzer — APPROVED** [review]
We checked the resent changes again, and tested. Everything works as it should, except the programming exercise results, which are not received on TS1. On TS3 this worked fine so it's not related to the PR.

**Review by fde312 — APPROVED** [review]
Tested with @SimonEntholzer and everything works fine. Code looks also good! 👍 Experienced the same issue already mentioned by Simon on TS1 but worked fine for TS3.

**Pull request merged**